### PR TITLE
Decorator improvements

### PR DESCRIPTION
**Pull request by elijahbenizzy**
A few decorator changes. Some are user-facing and some are relevant till later.

## Changes
1. The `target` parameter for NodeTransformer/NodeDecorator
2. Ensuring decorators declare their required configurations
3. Documentation improvements

## How I tested this

Automated tests.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: decorators.md [hunk 1]
```diff
@@ -558,3 +568,81 @@ Models (optionally) accept a `output_column` parameter -- this is specifically i
...
 it all later. At Stitch Fix this is necessary because various dependent columns that a model queries
 (e.g. `MULTIPLIER_...` and `OFFSET_...`) are derived from the model's name.
+
+## A mental model for decorators
+
+Decorators fall under a few broad categories:
+
+1. Decorators that decide whether a function should resolves into a node (E.G. config.when())
+2. Decorators that *turn* a function into a set of nodes (E.G. `@does`)
+3. Decorators that modify a set of nodes, turning it into another set of nodes (E.G. `@parameterize`, `@extract_columns`, and `@check_output`)
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply decorators.md:578]: There is a difference. The decorators in (3) *don't* operate on functions, they operate on nodes. 

The decorators in (2) have access to the function and use it to determine the subdag shape, the decorators in (2) don't.

File: decorators.md [hunk 2]
```diff
@@ -302,7 +304,13 @@ def intermediate_column() -> pd.Series:
...
     pass
 ```
-### @tag_outputs
+
+`@tag` also allows you to specify a target with the `target_` parameter. See "A mental model for decorators" below for more details.
+
+
+### @tag_outputs (deprecated)
+
+NOTE: tag_outputs is deprecated. Please use `@tag` instead with the `target_` parameter.
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply decorators.md:313]: Fair point -- I've changed my mind and will un-deprecate. I'll just make it use the same approach so the code is simpler.

File: decorators.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (skrawcz) [reply decorators.md:576]: Yes, but does that explanation make sense to an end user? I don't think we've articulated the distinction...

File: hamilton/function_modifiers/base.py [hunk 3]
```diff
@@ -233,30 +268,93 @@ def allows_multiple(cls) -> bool:
...
+    ) -> Collection[node.Node]:
+        """Given a set of nodes, and a set of nodes to transform, returns the set of nodes that
+        are not in the set of nodes to transform.
+
+        :param all_nodes: All nodes in the subdag
+        :param nodes_to_transform: All nodes to transform
+        :return: A collection of nodes that are not in the set of nodes to transform but are in the
+        subdag
+        """
+        return [node_ for node_ in all_nodes if node_ not in nodes_to_transform]
```
Comment (skrawcz) [inline hamilton/function_modifiers/base.py:357]: sure -- could also use a set operation here.

**Review by skrawcz — APPROVED** [review]
I haven't tested this -- seems to work.

File: decorators.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply decorators.md:599]: Yeah, TBH I think we need to work on making it clearer, and its difficult for me to do when I'm in the nodes headspace. So, let's revisit that and see how users react? Added a note about functions/nodes, but let's see how users react when we're rewriting documentation.

**Pull request merged**